pkg/cloudprovider/vsphere/nodemanager.go

```diff
@@ -118,195 +103,50 @@ func (nm *NodeManager) removeNode(node *v1.Node) {
-		return vclib.ErrNoVMFound
-	}
-	type vmSearch struct {
-		vc         string
-		datacenter *vclib.Datacenter
-	}
-
-	var mutex = &sync.Mutex{}
-	var globalErrMutex = &sync.Mutex{}
-	var queueChannel chan *vmSearch
-	var wg sync.WaitGroup
-	var globalErr *error
-
-	queueChannel = make(chan *vmSearch, cm.QUEUE_SIZE)
-
-	myNodeID := nodeID
-	if searchBy == FindVMByUUID {
-		glog.V(3).Info("DiscoverNode by UUID")
-		myNodeID = strings.ToLower(nodeID)
-	} else {
-		glog.V(3).Info("DiscoverNode by Name")
-	}
-	glog.V(2).Info("DiscoverNode nodeID: ", myNodeID)
-
-	vmFound := false
-	globalErr = nil
-
-	setGlobalErr := func(err error) {
-		globalErrMutex.Lock()
-		globalErr = &err
-		globalErrMutex.Unlock()
-	}
+	ctx := context.Background()
 
-	setVMFound := func(found bool) {
-		mutex.Lock()
-		vmFound = found
-		mutex.Unlock()
+	vmDI, err := nm.connectionManager.WhichVCandDCByNodeId(ctx, nodeID, cm.FindVM(searchBy))
+	if err != nil {
+		glog.Errorf("WhichVCandDCByNodeId failed. Err: %v", err)
 	}
 
-	getVMFound := func() bool {
-		mutex.Lock()
-		found := vmFound
-		mutex.Unlock()
-		return found
+	var oVM mo.VirtualMachine
+	err = vmDI.VM.Properties(ctx, vmDI.VM.Reference(), []string{"guest"}, &oVM)
+	if err != nil {
+		glog.Errorf("Error collecting properties for vm=%+v in vc=%s and datacenter=%s: %v",
+			vmDI.VM, vmDI.VcServer, vmDI.DataCenter.Name(), err)
+		return err
 	}
 
-	ctx := context.Background()
-
-	go func() {
-		for vc, vsi := range nm.connectionManager.VsphereInstanceMap {
-			var datacenterObjs []*vclib.Datacenter
-
-			found := getVMFound()
-			if found == true {
-				break
-			}
-
-			var err error
-			for i := 0; i < 3; i++ {
-				err = nm.connectionManager.Connect(ctx, vc)
-				if err == nil {
-					break
-				}
-				time.Sleep(1 * time.Second)
-			}
-
-			if err != nil {
-				glog.Error("Discovering node error vc:", err)
-				setGlobalErr(err)
-				continue
-			}
-
-			if vsi.Cfg.Datacenters == "" {
-				datacenterObjs, err = vclib.GetAllDatacenter(ctx, vsi.Conn)
-				if err != nil {
-					glog.Error("Discovering node error dc:", err)
-					setGlobalErr(err)
-					continue
-				}
-			} else {
-				datacenters := strings.Split(vsi.Cfg.Datacenters, ",")
-				for _, dc := range datacenters {
-					dc = strings.TrimSpace(dc)
-					if dc == "" {
-						continue
-					}
-					datacenterObj, err := vclib.GetDatacenter(ctx, vsi.Conn, dc)
-					if err != nil {
-						glog.Error("Discovering node error dc:", err)
-						setGlobalErr(err)
-						continue
-					}
-					datacenterObjs = append(datacenterObjs, datacenterObj)
-				}
-			}
-
-			for _, datacenterObj := range datacenterObjs {
-				found := getVMFound()
-				if found == true {
-					break
-				}
-
-				glog.V(4).Infof("Finding node %s in vc=%s and datacenter=%s", myNodeID, vc, datacenterObj.Name())
-				queueChannel <- &vmSearch{
-					vc:         vc,
-					datacenter: datacenterObj,
-				}
+	addrs := []v1.NodeAddress{}
+	for _, v := range oVM.Guest.Net {
```

Comment:
Consider is using this method, which only returns IP address that have a corresponding vSphere virtual NIC device: https://github.com/vmware/govmomi/blob/e560c7bb1e709826a897c9da4baad0944c2cdda0/object/virtual_machine.go#L254-L259

Also see discussion in: https://github.com/kubernetes/kubernetes/pull/73721